atomsci/modac/__init__.py

```diff
@@ -136,3 +165,46 @@ def download_all_files_in_collection(
                     if sub_collection_path:
                         self.download_all_files_in_collection(sub_collection_path, True)
 
+    @ensure_authenticated
+    @log_action
+    def upload_file(
+        self,
+        collection_path: str,
+        data_file_path: str,
+        attributes_file: Optional[Dict] = None,
+    ) -> bool:
+        """
+        Upload a file to the MoDaC system from local filesystem.
+
+        :param collection_path: Path where the data file should be registered in MoDaC.
+        :param data_file_path: Path to the data file on the local filesystem.
+        :param attributes_file: JSON-like dictionary containing metadata about the data file.
+        """
+        url = "/".join((self.BASE_URL, "v2", "dataObject", collection_path))
+
+        if attributes_file is None:
+            attributes_file = {}
+
+        files = {
+            "dataObjectRegistration": (
+                "attributesFile.json",
+                json.dumps(attributes_file),
+                "application/json",
+            ),
+            "dataObject": (
+                os.path.basename(data_file_path),
+                open(data_file_path, "rb"),
+            ),
+        }
+
+        _logger.warning(f"Uploading file to {url}")
+        try:
+            with requests.put(url, files=files, headers=self._token_headers()) as resp:
+                resp.raise_for_status()
+                _logger.info(f"File uploaded successfully: {resp.json()}")
+                return True
+        except requests.exceptions.RequestException as e:
+            _logger.error(f"File upload failed: {e}")
+            raise
```

Comment:
```suggestion
            raise e
```
It's good practice to raise the previous exception unless you have a custom one to raise

We could make our own custom exception for this module but that might not be necessary yet